### PR TITLE
certbot: add --preferred-chain

### PR DESCRIPTION
**Pull request by alexzorin**
For #6971.

@bmw no tests for the Certbot bit yet, but I'd like to wait to see whether you have any high-level comments before I go any deeper.

### CODE REVIEW

**Comment by bmw** [comment]
Thanks @alexzorin.

I'll make sure we take a look at this soon.

File: certbot/certbot/interfaces.py [hunk 1]
```diff
@@ -254,6 +254,13 @@ class IConfig(zope.interface.Interface):
         "If updates provided by installer enhancements when Certbot is being run"
         " with \"renew\" verb should be disabled.")
 
+    preferred_chain = zope.interface.Attribute(
+        "If the CA offers multiple certificate chains for an issued certificate"
+        ", prefer the chain with an issuer matching this Subject Common Name or "
+        "Subject Key Identifier (hex-encoded). If no match, the default offered "
```
[the remark this answers is not in the document]
Comment (alexzorin) [reply certbot/certbot/interfaces.py:260]: You are right, it is borne out of my own misunderstanding of what's possible under path validation. I will reduce it to only issuer CN.

File: certbot/certbot/crypto_util.py [hunk 2]
```diff
@@ -530,3 +531,47 @@ def get_serial_from_cert(cert_path):
...
+                aki_ext = cert.extensions.get_extension_for_class(x509.AuthorityKeyIdentifier)
+                if aki_ext and aki_ext.value.key_identifier == ski_bytes:
+                    return chain
+            except x509.ExtensionNotFound:
+                # The AKI extension should always be present for any CA, but we shouldn't
+                # raise an error if it isn't.
+                pass
+
+    # Nothing matched, return whatever was first in the list.
+    return fullchains[0]
```
[the remark this answers is not in the document]
Comment (alexzorin) [reply certbot/certbot/crypto_util.py:577]: (So I remember: and if it's not a dry-run). 

I see an argument for the warning: inform the user their flag is not/no longer effective.

It is however a time bomb: cron emails when a CA no longer uses that issuer. The user would have to then figure out how to identify and remove the flag from their renewal configs.

What do you think? Given the "hint" nature of the flag, I'm not sure it's warranted to bother the user about it. One compromise could be to emit it only on interactive renewals.

[the remark this answers is not in the document]
Comment (bmw) [reply certbot/certbot/crypto_util.py:561]: That's a good point. How about we change the call from `logger.warning` to `logger.info`? With this change, it'll be printed to the terminal only if `-q` or `--quiet` isn't included on the command line.

**Comment by bmw** [comment]
Another thing I think we need to do here is make sure this value is saved for renewal. Accomplishing this should be as easy as adding `preferred_chain` to the top of `STR_CONFIG_ITEMS` in `certbot/certbot/_internal/renewal.py`

**Comment by alexzorin** [comment]
>pkg_resources.DistributionNotFound: acme>=1.6.0.dev0

Hmm. I tried following the advice (https://github.com/certbot/certbot/pull/7266#issuecomment-515500053) from last time I had to bump the `acme` dependency, and I didn't touch anything since last time this test was passing ...

**Comment by bmw** [comment]
Ah sorry. It's a transitive dependency problem. `certbot-nginx`'s tests here is trying to use`acme==1.4.0` and local, unreleased Certbot which isn't going to work because with this PR unreleased Certbot requires the unreleased version of ACME.

To fix this, I'd update `certbot-nginx`'s `setup.py` and `local-oldest-requirements.txt` the same way you did for Certbot.

**Comment by bmw** [comment]
And the reason this isn't a problem for the Apache plugin with a similar setup right now is because apparently in Apache's unit tests we avoid this `pkg_resources` code that checks that dependencies are satisfied 🤦‍♂️  In addition to the unit tests, we're running some system tests with the nginx plugin here.

**Pull request merged**